x/mongo/driver/topology/server.go

```diff
@@ -707,28 +725,34 @@ func (s *Server) check() (description.Server, error) {
 	var err error
 	var durationNanos int64
 
+	previousDescription := s.Description()
+	streamable := previousDescription.TopologyVersion != nil
+
+	isNilConn := s.conn == nil
+	if !isNilConn {
+		s.publishServerHeartbeatStartedEvent(s.conn.ID(), s.conn.getCurrentlyStreaming() || streamable)
+	}
 	// Create a new connection if this is the first check, the connection was closed after an error during the previous
 	// check, or the previous check was cancelled.
-	if s.conn == nil || s.conn.closed() || s.checkWasCancelled() {
+	if isNilConn || s.conn.closed() || s.checkWasCancelled() {
 		// Create a new connection and add it's handshake RTT as a sample.
 		err = s.setupHeartbeatConnection()
 		if err == nil {
 			// Use the description from the connection handshake as the value for this check.
 			s.rttMonitor.addSample(s.conn.helloRTT)
 			descPtr = &s.conn.desc
+		} else {
+			err = unwrapConnectionError(err)
 		}
 	}
 
-	if descPtr == nil && err == nil {
+	if !isNilConn && err == nil {
```

Comment:
I think these are slightly different checks. `s.conn.desc == nil` vs. `s.conn != nil` IIUC. What's the reasoning for the change?

Reply:
I suppose I'm confused why you're using `!isNilConn`, then. The way I understand it, we want to set up a heartbeat connection if `s.conn == nil` on [L744](https://github.com/mongodb/mongo-go-driver/pull/1133/files#diff-5e467b910a8117877d35340d0673292dbe5435fb482aa701363a22984daec72dR744); once we've done that, assuming there was no error, `s.conn != nil`. Should we run _another_ heartbeat below if `s.conn != nil`?

Reply:
Discussed offline; the logic seems correct to me. My misunderstanding was that `isNilConn` does not represent whether `s.conn` is `nil` at any given moment, it represents whether `s.conn` _was_ `nil` at L738. Can we leave a comment above this line with something like

```
// If !isNilConn, s.conn was not nil at the start of check() and we can use the existing connection for the heartbeat.
```

Reply:
@benjirewis that comment is not always true. For example, if `s.conn` is not `nil`, but is closed, the `if isNilConn || s.conn.closed() || ...` block above will create a new connection and set the server description based on the response from the new connection handshake. In that case, the updated code would create a new connection and then immediately start a new heartbeat.

The intent of the `if descPtr == nil && err == nil` check is to avoid immediately running another heartbeat after creating a connection, which the `if !isNilConn && err == nil` check won't always do. I think we should revert it to the previous condition.

Does this seem like correct logic?
```go
if s.conn == nil || s.conn.closed() || s.checkWasCancelled() {
	// Create a new connection and record the server description from the handshake.
} else {
	// Run the next heartbeat with the existing connection.
}
```

Reply:
@matthewdale I think you are correct. I was just wondering why the code wasn't in this way originally. 🤔

Reply:
Thank you both for the clarifications here and offline; the simpler `if else` logic makes sense to me.